### PR TITLE
BackYourStack dispatcher

### PR DESCRIPTION
**Pull request by znarf**
Continuing from https://github.com/opencollective/opencollective-api/pull/2153

### CODE REVIEW

File: server/lib/backyourstack/dispatcher.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+    type: 'prepaid',
+    uuid: uuidV4(),
+    data: { HostCollectiveId },
+  });
+
+  return map(
+    depRecommendation,
+    async dependency => {
+      // Check if the collective is avaliable
+      const collective = await models.Collective.findOne({ where: { slug: dependency.opencollective.slug } });
```
Comment (znarf) [inline server/lib/backyourstack/dispatcher.js]: @flickz this should be refactored so we filter existing collectives before. The goal is to compute `sumOfWeights` only after the collectives are filtered.

Comment (flickz) [reply server/lib/backyourstack/dispatcher.js]: I've updated this now

File: server/lib/backyourstack/dispatcher.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    debug('Error fetching dependcies', err);
+    console.error(err);
+    throw new Error('Unable to fetch dependencies, please ensure the url is correct');
+  }
+
+  for (const depRecommended of depRecommendations) {
+    const collective = await models.Collective.findOne({ where: { slug: depRecommended.opencollective.slug } });
+    if (!collective) {
+      console.log(`Unable to fetch collective with slug ${depRecommended.opencollective.slug}`);
+      return;
```
Comment (znarf) [inline server/lib/backyourstack/dispatcher.js]: Have you rely tried that?

File: server/lib/backyourstack/dispatcher.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    }
+    collective.totalAmount = computeAmount(shareableAmount, sumOfWeights, depRecommended.weight);
```
Comment (znarf) [inline server/lib/backyourstack/dispatcher.js]: Is `sumOfWeights` defined at this point?

File: server/lib/backyourstack/dispatcher.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        CreatedByUserId: order.CreatedByUserId,
+        FromCollectiveId: order.FromCollectiveId,
+        CollectiveId: collective.id,
+        quantity: order.quantity,
+        description: `Monthly donation to ${collective.name} through BackYourStack`,
+        totalAmount: collective.totalAmount,
+        currency: order.currency,
+        status: status.PENDING,
+      };
+      delete collective.totalAmount;
```
Comment (znarf) [inline server/lib/backyourstack/dispatcher.js]: What's the purpose of that?

File: server/paymentProviders/opencollective/prepaid.js [hunk 5]
```diff
@@ -80,9 +80,17 @@ async function processOrder(order) {
     throw new Error("This payment method doesn't have enough funds to complete this order");
   }
 
+  let hostFeeInHostCurrency, platformFeeInHostCurrency;
+
+  if (options.skipPlatformFee && options.skipHostFee) {
```
Comment (znarf) [inline server/paymentProviders/opencollective/prepaid.js]: That doesn't make sense. You should handle these properties individually.

File: server/lib/backyourstack/dispatcher.js (hunk 2 above)
Comment (znarf) [reply server/lib/backyourstack/dispatcher.js]: This needs to be addressed.

File: server/paymentProviders/opencollective/prepaid.js (hunk 5 above)
Comment (flickz) [reply server/paymentProviders/opencollective/prepaid.js]: Oh thought they were together. I'll change it

**Pull request merged**